xdsl/dialects/stencil.py

```diff
@@ -289,6 +290,50 @@ def __init__(self, float_t: AnyFloat) -> None:
         super().__init__([float_t])
 
 
+@irdl_op_definition
+class ApplyOp(IRDLOperation):
+    """
+    This operation takes a stencil function plus parameters and applies
+    the stencil function to the output temp.
+
+    Example:
+
+      %0 = stencil.apply (%arg0=%0 : !stencil.temp<?x?x?xf64>) -> !stencil.temp<?x?x?xf64> {
+        ...
+      }
+    """
+
+    name = "stencil.apply"
+    args: Annotated[VarOperand, Attribute]
+    region: Region
+    res: Annotated[VarOpResult, TempType]
+
+    @staticmethod
+    def get(
+        args: Sequence[SSAValue] | Sequence[Operation],
+        body: Block,
+        result_types: Sequence[TempType[Attribute]],
+    ):
+        assert len(result_types) > 0
+
+        return ApplyOp.build(
+            operands=[list(args)],
+            regions=[Region(body)],
+            result_types=[result_types],
+        )
+
+    def verify_(self) -> None:
+        if len(self.res) < 1:
+            raise VerifyException(
+                f"Expected stencil.apply to have at least 1 result, got {len(self.res)}"
+            )
+
+    def get_rank(self) -> int:
+        res_typ = self.res[0].typ
+        assert isa(res_typ, TempType[Attribute])
+        return res_typ.get_num_dims()
```

Comment:
If you intent for the rank to be uniquely identified, then you should probably verify that the rank of all results is the same, otherwise this seems unsound

Reply:
(and the input temps should also share that rank)

Reply:
Yup, I'm going in bites, but indeed for the result.
I don't think it's necessary for the inputs though: it's only needed if they are stencil.accessed, and this is checked.

Reply:
Do you have a plan on testing this?

Reply:
How are you going to get the rank?

Reply:
> How are you going to get the rank?

What do you mean? It's this method; it is the rank of the result(s). Yes, this is not verified to be fully coherent yet, as half the dialect, but a bit less than before with this PR; a following PR will handle this

> Do you have a plan on testing this?

Yes